SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,106 @@
+    return new ArrayList<>(activeTasksForRequest.values());
+  }
+
+  private Collection<UpstreamInfo> getUpstreamsFromActiveTasksForRequest(String requestId, Optional<String> loadBalancerUpstreamGroup) {
+    return lbClient.getUpstreamsForTasks(getActiveTasksForRequest(requestId), requestId, loadBalancerUpstreamGroup);
+  }
+
+  private boolean isEqualUpstreamGroupRackId(UpstreamInfo upstream1, UpstreamInfo upstream2){
+    return (upstream1.getUpstream().equals(upstream2.getUpstream()))
+        && (upstream1.getGroup().equals(upstream2.getGroup()))
+        && (upstream1.getRackId().equals(upstream2.getRackId()));
+  }
+
+  private Collection<UpstreamInfo> getEqualUpstreams(UpstreamInfo upstream, Collection<UpstreamInfo> upstreams) {
+    return upstreams.stream().filter(candidate -> isEqualUpstreamGroupRackId(candidate, upstream)).collect(Collectors.toList());
+  }
+
+  // TODO: confirm if there can only be one match, or it's fine to collect a list of matches as is right now
+  private List<UpstreamInfo> getExtraUpstreams(Collection<UpstreamInfo> upstreamsInBaragonForRequest, Collection<UpstreamInfo> upstreamsInSingularityForRequest) {
+    for (UpstreamInfo upstreamInSingularity : upstreamsInSingularityForRequest) {
+      final Collection<UpstreamInfo> matches = getEqualUpstreams(upstreamInSingularity, upstreamsInBaragonForRequest);
+      upstreamsInBaragonForRequest.removeAll(matches);
+    }
+    return new ArrayList<>(upstreamsInBaragonForRequest);
+  }
+
+  private SingularityLoadBalancerUpdate syncUpstreamsForService(SingularityRequest singularityRequest, SingularityDeploy deploy, Optional<String> loadBalancerUpstreamGroup) {
+    final String singularityRequestId = singularityRequest.getId();
+    final LoadBalancerRequestId loadBalancerRequestId = new LoadBalancerRequestId(singularityRequestId, LoadBalancerRequestType.REMOVE, Optional.absent());
+    Collection<UpstreamInfo> upstreamsInBaragonForRequest = lbClient.getBaragonUpstreamsForRequest(loadBalancerRequestId.toString());
+    Collection<UpstreamInfo> upstreamsInSingularityForRequest = getUpstreamsFromActiveTasksForRequest(singularityRequestId, loadBalancerUpstreamGroup);
+    final List<UpstreamInfo> extraUpstreams = getExtraUpstreams(upstreamsInBaragonForRequest, upstreamsInSingularityForRequest);
+
+    return lbClient.makeAndSendBaragonRequest(loadBalancerRequestId, Collections.emptyList(), extraUpstreams, deploy, singularityRequest);
+  }
+
+  public void syncUpstreams() {
+    for (SingularityRequestWithState singularityRequestWithState: requestManager.getActiveRequests()){
+      final SingularityRequest singularityRequest= singularityRequestWithState.getRequest();
+      if (singularityRequest.isLoadBalanced()) {
+        final String singularityRequestId = singularityRequest.getId(); //TODO: lock on the requestId
+        final Optional<String> maybeDeployId = deployManager.getInUseDeployId(singularityRequestId);
```

Comment:
additional thing. I'm wondering if we should skip the sync if there is a deploy in progress. I'm trying to think through all the ways this could possibly corrupt state, and I could see a deploy and bounce being problematic. See my comment on `getActiveTasksForRequest` for more detail